test/InterpolateTests.hs

```diff
@@ -23,16 +31,149 @@ tests =
+              , "</html>"
+              ]
+        result @?= expected
+    , testCase "interpolates SQL" $ do
+        let
+          name = "Alice"
+          age = 30
+          active = False
+          sqlInjection = "TRUE; DELETE FROM user"
+          sqlSafe = "(foo = 'bar' OR bar = 'foo')" :: SqlQuery
+          result =
+            s"""
+            SELECT * FROM user
+            WHERE
+              name = ${name} AND
+              age = ${age} AND
+              active = ${active} AND
+              failedInjection = ${sqlInjection} AND
+              ${sqlSafe}
+            """
+          expected =
+            SqlQuery
+              ( Text.unlines
+                  [ "SELECT * FROM user"
+                  , "WHERE"
+                  , "  name = ? AND"
+                  , "  age = ? AND"
+                  , "  active = ? AND"
+                  , "  failedInjection = ? AND"
+                  , "  (foo = 'bar' OR bar = 'foo')"
+                  ]
+              )
+              [ SqlString name
+              , SqlInt age
+              , SqlBool active
+              , SqlString sqlInjection
+              ]
+        result @?= expected
     ]
 
--- TODO: HTML interpolation
--- TODO: SqlQuery interpolation
+{----- HTML interpolation -----}
+
+data HTML = HTML Text
+  deriving (Show, Eq)
+
+instance IsString HTML where
+  fromString = HTML . Text.pack
+
+instance Interpolate HTML where
+  interpolateRaw s (HTML t) = HTML (Text.pack s <> t)
+  interpolateEmpty = HTML ""
+
+instance InterpolateValue HTML HTML where
+  interpolatePrec _ (HTML s) (HTML t) = HTML (s <> t)
+
+instance {-# OVERLAPPING #-} InterpolateValue HTML String where
+  interpolatePrec _ s (HTML t) = HTML $ escape (Text.pack s) <> t
+    where
+      escape = Text.replace "<" "&lt;" . Text.replace ">" "&gt;"
+
+{----- SqlQuery interpolation -----}
+
+data SqlValue
+  = SqlString Text
+  | SqlInt Int
+  | SqlBool Bool
+  | SqlList [SqlValue]
+  deriving (Show, Eq)
+
+class ToSqlValue a where
+  toSqlValue :: a -> SqlValue
+instance {-# OVERLAPPING #-} ToSqlValue String where
+  toSqlValue = SqlString . Text.pack
+instance ToSqlValue Text where
+  toSqlValue = SqlString
+instance ToSqlValue Int where
+  toSqlValue = SqlInt
+instance ToSqlValue Bool where
+  toSqlValue = SqlBool
+instance ToSqlValue a => ToSqlValue [a] where
+  toSqlValue = SqlList . map toSqlValue
+
+data SqlQuery = SqlQuery Text [SqlValue]
+  deriving (Show, Eq)
+
+instance IsString SqlQuery where
+  fromString s = SqlQuery (Text.pack s) []
+instance Semigroup SqlQuery where
+  SqlQuery s vs <> SqlQuery s' vs' = SqlQuery (s <> s') (vs <> vs')
+instance Monoid SqlQuery where
+  mempty = SqlQuery "" []
+
+instance Interpolate SqlQuery
+
+instance {-# OVERLAPPING #-} InterpolateValue SqlQuery SqlQuery where
+  interpolatePrec _ = (<>)
+
+-- instance {-# OVERLAPPING #-} ToSqlValue a => InterpolateValue SqlQuery a where
+--   interpolate a = SqlQuery "?" [toSqlValue a]
```

Comment:
Trying to uncomment this. The problem is that GHC is unable to figure out which instance to pick; e.g. when resolving `InterpolateValue SqlQuery Bool`, which of the following should be picked:

```hs
-- the default one
instance InterpolateValue s String => InterpolateValue s Bool where
  interpolate = interpolate . show

-- this one
instance ToSqlValue a => InterpolateValue SqlQuery a where
  interpolate a = SqlQuery "?" [toSqlValue a]
```

Neither is more specific than the other.

Trying one solution here: 29aebd88a9cbfc1db0538cab041266e63a60d11e, using the auxiliary class method described in https://wiki.haskell.org/GHC/AdvancedOverlap, but this still gives me overlapping instances:

```hs
-- resolving
InterpolateValueFlag [Char] [Char] InterpolateOverride

-- found
instance (flag ~ InterpolateDefault) => InterpolateValueFlag s a flag
instance InterpolateValueFlag String String InterpolateOverride
```

Why isn't the second instance being chosen? It looks like the exact constraint being asked for. I also don't see how the `ShowPred` example works, if this isn't working

Reply:
Resolved: https://github.com/brandonchinn178/string-syntax/pull/2/commits/9f15664b98dd0cf8d2124322eac0db3de90716f8